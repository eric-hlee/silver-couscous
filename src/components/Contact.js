import React, { useState } from 'react'
import Button from 'react-bootstrap/Button'
import Col from 'react-bootstrap/Col'
import Form from 'react-bootstrap/Form'

export default function Contact() {
	const [validated, setValidated] = useState(false)

	const handleSubmit = (event) => {
		const form = event.currentTarget
		if (form.checkValidity() === false) {
			event.preventDefault()
			event.stopPropagation()
		}

		setValidated(true)
	}

	return (
		<section className="contact-form">
			<h1 className="contact">Contact me</h1>
			<Form
				name="contact"
				method="post"
				className="contact"
				noValidate
				validated={validated}
				onSubmit={handleSubmit}
			>
				<Form.Group controlId="formName&Email">
					<Form.Row>
						<Col>
							<Form.Label>Name</Form.Label>
							<Form.Control
								type="text"
								name="name"
								placeholder="Name"
								required
							/>
							<Form.Control.Feedback type="invalid">
								Please enter your name.
							</Form.Control.Feedback>
						</Col>
						<Col>
							<Form.Label>Email address</Form.Label>
							<Form.Control
								type="email"
								name="email"
								placeholder="Enter email"
								required
							/>
							<Form.Control.Feedback type="invalid">
								Please provide a valid email.
							</Form.Control.Feedback>
						</Col>
					</Form.Row>
				</Form.Group>
				<Form.Group controlId="formMessage">
					<Form.Label>Message</Form.Label>
					<Form.Control as="textarea" name="message" rows="3" required />
					<Form.Control.Feedback type="invalid">
						Please enter a message.
					</Form.Control.Feedback>
				</Form.Group>
				<Button className="submit-button" variant="dark" type="submit">
					Submit
				</Button>
			</Form>
		</section>
	)
}
