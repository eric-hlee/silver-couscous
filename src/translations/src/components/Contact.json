[{"id":"contact.contact_me","defaultMessage":"Contact me","file":"src/components/Contact.js","start":{"line":78,"column":8},"end":{"line":78,"column":80}},{"id":"contact.name","defaultMessage":"Name","file":"src/components/Contact.js","start":{"line":92,"column":16},"end":{"line":92,"column":76}},{"id":"contact.invalid_name","defaultMessage":"Please enter your name.","file":"src/components/Contact.js","start":{"line":103,"column":16},"end":{"line":106,"column":18}},{"id":"contact.email","defaultMessage":"Email address","file":"src/components/Contact.js","start":{"line":111,"column":16},"end":{"line":114,"column":18}},{"id":"contact.invalid_email","defaultMessage":"Please provide a valid email.","file":"src/components/Contact.js","start":{"line":125,"column":16},"end":{"line":128,"column":18}},{"id":"contact.message","defaultMessage":"Message","file":"src/components/Contact.js","start":{"line":135,"column":12},"end":{"line":135,"column":78}},{"id":"contact.invalid_message","defaultMessage":"Please enter a message.","file":"src/components/Contact.js","start":{"line":146,"column":12},"end":{"line":149,"column":14}},{"id":"contact.submit","defaultMessage":"Submit","file":"src/components/Contact.js","start":{"line":153,"column":10},"end":{"line":153,"column":74}},{"id":"contact.success","defaultMessage":"🎉 Success!","file":"src/components/Contact.js","start":{"line":159,"column":12},"end":{"line":162,"column":14}},{"id":"contact.success_message","defaultMessage":"If necessary, I will respond to you as soon as possible.","file":"src/components/Contact.js","start":{"line":167,"column":12},"end":{"line":170,"column":14}},{"id":"contact.close","defaultMessage":"Close","file":"src/components/Contact.js","start":{"line":175,"column":12},"end":{"line":175,"column":74}}]
