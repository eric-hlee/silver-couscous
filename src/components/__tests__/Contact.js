import React from 'react'
import { act } from 'react-dom/test-utils'
import { shallow, mount } from 'enzyme'
import { IntlProvider } from 'react-intl'
import Contact from '../Contact'

describe('<Contact />', () => {
	const getComponent = () =>
		mount(
			<IntlProvider locale="en">
				<Contact />
			</IntlProvider>,
			{ attachTo: window.domNode }
		)
	let div
	beforeEach(() => {
		div = document.createElement('div')
		window.domNode = div
		document.body.appendChild(div)
		jest.spyOn(global, 'fetch').mockImplementation(() => {
			return Promise.resolve({
				json: () => Promise.resolve({}),
			})
		})
	})

	afterEach(() => {
		document.body.removeChild(div)
		global.fetch.mockRestore()
	})

	it('renders correctly', () => {
		const wrapper = shallow(<Contact />)
		expect(wrapper).toMatchSnapshot()
	})

	it('posts to / and shows modal on form submission success', async () => {
		const wrapper = getComponent()

		const nameInput = wrapper.find("input[name='name']")
		nameInput.simulate('change', { target: { name: 'name', value: 'eric' } })

		const emailInput = wrapper.find("input[name='email']")
		emailInput.simulate('change', {
			target: { name: 'email', value: 'eric@test.com' },
		})

		const textarea = wrapper.find('textarea')
		textarea.simulate('change', {
			target: { name: 'message', value: 'testing' },
		})

		const submitButton = wrapper.find("Button[type='submit']")
		await act(async () => {
			submitButton.simulate('submit')
		})

		expect(global.fetch).toHaveBeenCalledTimes(1)
		expect(global.fetch).toHaveBeenCalledWith('/', {
			body: 'form-name=contact&name=eric&email=eric%40test.com&message=testing',
			headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
			method: 'POST',
		})

		wrapper.update()
		expect(wrapper.find('Modal').first().prop('show')).toBe(true)
	})

	it('shows feedback message and does not post to / on input validation failure', async () => {
		const wrapper = getComponent()

		const nameInput = wrapper.find("input[name='name']")
		nameInput.simulate('change', { target: { name: 'name', value: '' } })

		const emailInput = wrapper.find("input[name='email']")
		emailInput.simulate('change', {
			target: { name: 'email', value: '' },
		})

		const textarea = wrapper.find('textarea')
		textarea.simulate('change', {
			target: { name: 'message', value: '' },
		})

		const submitButton = wrapper.find("Button[type='submit']")
		await act(async () => {
			submitButton.simulate('submit')
		})

		wrapper.update()
		expect(wrapper.find('Feedback[type="invalid"]').first().text()).toEqual(
			'Please enter your name.'
		)
		expect(wrapper.find('Feedback[type="invalid"]').at(1).text()).toEqual(
			'Please provide a valid email.'
		)
		expect(wrapper.find('Feedback[type="invalid"]').last().text()).toEqual(
			'Please enter a message.'
		)
		expect(global.fetch).not.toHaveBeenCalled()
	})

	it('hides modal on close button click', async () => {
		const wrapper = getComponent()

		const nameInput = wrapper.find("input[name='name']")
		nameInput.simulate('change', { target: { name: 'name', value: 'eric' } })

		const emailInput = wrapper.find("input[name='email']")
		emailInput.simulate('change', {
			target: { name: 'email', value: 'eric@test.com' },
		})

		const textarea = wrapper.find('textarea')
		textarea.simulate('change', {
			target: { name: 'message', value: 'testing' },
		})

		const submitButton = wrapper.find("Button[type='submit']")
		await act(async () => {
			submitButton.simulate('submit')
		})

		wrapper.update()
		const closeButton = wrapper.find("Button[variant='secondary']")
		closeButton.simulate('click')
		expect(wrapper.find('Modal').first().prop('show')).toBe(false)
	})
})
