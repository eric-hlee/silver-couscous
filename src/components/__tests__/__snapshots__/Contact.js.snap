// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Contact /> renders correctly 1`] = `
<section
  className="contact-form"
>
  <h1
    className="contact"
  >
    <FormattedMessage
      defaultMessage="Contact me"
      id="contact.contact_me"
    />
  </h1>
  <Form
    className="contact"
    inline={false}
    method="post"
    name="contact"
    noValidate={true}
    onSubmit={[Function]}
    validated={false}
  >
    <FormGroup
      controlId="formNameEmail"
    >
      <FormRow>
        <Col>
          <FormLabel
            column={false}
            srOnly={false}
          >
            <FormattedMessage
              defaultMessage="Name"
              id="contact.name"
            />
          </FormLabel>
          <FormControl
            name="name"
            onChange={[Function]}
            placeholder="Name"
            required={true}
            type="text"
            value=""
          />
          <Feedback
            type="invalid"
          >
            <FormattedMessage
              defaultMessage="Please enter your name."
              id="contact.invalid_name"
            />
          </Feedback>
        </Col>
        <Col>
          <FormLabel
            column={false}
            srOnly={false}
          >
            <FormattedMessage
              defaultMessage="Email address"
              id="contact.email"
            />
          </FormLabel>
          <FormControl
            name="email"
            onChange={[Function]}
            placeholder="Enter email"
            required={true}
            type="email"
            value=""
          />
          <Feedback
            type="invalid"
          >
            <FormattedMessage
              defaultMessage="Please provide a valid email."
              id="contact.invalid_email"
            />
          </Feedback>
        </Col>
      </FormRow>
    </FormGroup>
    <FormGroup
      controlId="formMessage"
    >
      <FormLabel
        column={false}
        srOnly={false}
      >
        <FormattedMessage
          defaultMessage="Message"
          id="contact.message"
        />
      </FormLabel>
      <FormControl
        as="textarea"
        name="message"
        onChange={[Function]}
        required={true}
        rows="3"
        value=""
      />
      <Feedback
        type="invalid"
      >
        <FormattedMessage
          defaultMessage="Please enter a message."
          id="contact.invalid_message"
        />
      </Feedback>
    </FormGroup>
    <Button
      active={false}
      className="submit-button"
      disabled={false}
      type="submit"
      variant="dark"
    >
      <FormattedMessage
        defaultMessage="Submit"
        id="contact.submit"
      />
    </Button>
  </Form>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    dialogAs={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "ModalDialog",
        "render": [Function],
      }
    }
    enforceFocus={true}
    keyboard={true}
    onHide={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle>
        <FormattedMessage
          defaultMessage="ðŸŽ‰ Success!"
          id="contact.success"
        />
      </ModalTitle>
    </ModalHeader>
    <ModalBody>
      <p>
        <FormattedMessage
          defaultMessage="If necessary, I will respond to you as soon as possible."
          id="contact.success_message"
        />
      </p>
    </ModalBody>
    <ModalFooter>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="secondary"
      >
        <FormattedMessage
          defaultMessage="Close"
          id="contact.close"
        />
      </Button>
    </ModalFooter>
  </Modal>
</section>
`;
